Salt Solution Code Review

alan-krumholz opened the pull request:
Found minor error in the README file of the energy solution and I'm also fixing that error as part of this commit.

Commiting changes to convert baseline solution into the saltt segmentation solution.

wes-turner commented on machine-learning/solutions/seismic_salt_detection/config.yaml:
```diff
@@ -45,9 +35,22 @@ trainingInput:
-      scaleType: UNIT_LOG_SCALE
\ No newline at end of file
+      scaleType: UNIT_LOG_SCALE
+    - parameterName: module
```
nit: "module" is unambiguous within the TF Hub framework.  Outside of it, the meaning is less clear.  What do you think of tf_hub_module as a name?

alan-krumholz replied:
fixed

wes-turner commented on machine-learning/solutions/seismic_salt_detection/setup.py:
```diff
@@ -21,12 +21,12 @@
     install_requires=REQUIRED_PACKAGES,
     packages=find_packages(),
-    description='This model forecasts hourly energy prices.'
+    description='This model detects salt on seismic image.'
```
nit: "detects salt in (a seismic image|seismic images)"

alan-krumholz replied:
will do

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -27,9 +27,7 @@
 
```
This file has a handful of lint errors from gpylint.  I'm just going to paste them in right here.  In the future, if you want to run gpylint yourself, the option for changing the Google-default 2-space indent (which gpylint uses by default) to PEP8 4-space is:
  indent-string='    '

You can either dump an rcfile with `gpylint --generate-rcfile` and edit it yourself, or you can use one I just copied to https://cs.corp.google.com/piper///depot/google3/experimental/users/westurner/style/gpylint.rc.  To use, `gpylint --rcfile=<rcfile> <file-to-lint>`

************* File trainer/inputs.py
C6016:  2: No space after start of comment [g-no-space-after-comment]
go/gpylint-faq#g-no-space-after-comment
C6016:  4: No space after start of comment [g-no-space-after-comment]
go/gpylint-faq#g-no-space-after-comment
C6016:  5: No space after start of comment [g-no-space-after-comment]
go/gpylint-faq#g-no-space-after-comment
C6016:  6: No space after start of comment [g-no-space-after-comment]
go/gpylint-faq#g-no-space-after-comment
C6016: 10: No space after start of comment [g-no-space-after-comment]
go/gpylint-faq#g-no-space-after-comment
C6016: 11: No space after start of comment [g-no-space-after-comment]
go/gpylint-faq#g-no-space-after-comment
C6016: 12: No space after start of comment [g-no-space-after-comment]
go/gpylint-faq#g-no-space-after-comment
C6016: 13: No space after start of comment [g-no-space-after-comment]
go/gpylint-faq#g-no-space-after-comment
C6016: 14: No space after start of comment [g-no-space-after-comment]
go/gpylint-faq#g-no-space-after-comment
C6020: 62: Need 2 blank lines before top level def [g-wrong-blank-lines]
go/gpylint-faq#g-wrong-blank-lines
C0326: 73: Exactly one space required after comma
    image_decoded = tf.image.decode_png(image_string,channels=3)
                                                    ^
 [bad-whitespace]
go/gpylint-faq#bad-whitespace
C0303: 75: Trailing whitespace [trailing-whitespace]
go/gpylint-faq#trailing-whitespace
C6020: 76: Need 2 blank lines before top level def [g-wrong-blank-lines]
go/gpylint-faq#g-wrong-blank-lines
C0301: 87: Line too long (87/80) [line-too-long]
go/gpylint-faq#line-too-long
C0326: 87: Exactly one space required after comma
    features['image'] = load_image(image_path + tf.reshape(features['id'],[]) + '.png')
                                                                         ^
 [bad-whitespace]
go/gpylint-faq#bad-whitespace
C6402:121:generate_input_fn._input_fn.<lambda>: 'lambda' may be too long [g-long-lambda]
go/gpylint-faq#g-long-lambda
C0326:123: Exactly one space required after comma
            target),num_parallel_calls=num_threads)
                   ^
 [bad-whitespace]
go/gpylint-faq#bad-whitespace
C6020:133: Need 2 blank lines before top level def [g-wrong-blank-lines]
go/gpylint-faq#g-wrong-blank-lines
C0303:166: Trailing whitespace [trailing-whitespace]
go/gpylint-faq#trailing-whitespace
C6113:192: Args section is missing, does not match arg list or is badly indented.

Expected: validation_path, image_path, batch_size

Found: training_path, image_path, batch_size

Missing: validation_path

Unexpected: training_path
 [g-doc-args]
go/gpylint-faq#g-doc-args

alan-krumholz replied:
When doing the previous solution (in which this one is based) we decided not to use gpylint (as it is not publicly available) so we used public pylint instead and the code was passing all those checks.

wes-turner replied:
It's totally ok to not use gpylint if you don't want to -- gpylint is just a tool.  But I think it's still a good idea for us to follow the Google style guide.  Is this something you disagree with?

alan-krumholz replied:
It'll be much better if we could have an rcfile that works with public pylint so we can open source it and recommend it to partners. I talked to Brian about that and he was planning to produce one but I'm not sure if it ever happened. I'll use yours with gpylint just to get through this code review but I don't think we should be using any tools that are not externally available to produce these solutions if we expect partners/clients to be as efficient as we are.

alan-krumholz replied:
will fix

wes-turner replied:
I totally agree.  b/29253510 for some context.

I think it'd be a fun project to work on, myself -- I've never made any substantial contributions to any widely-used open-source libraries.  It's just been hard so far for me to prioritize.

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -27,9 +27,7 @@
 
 import tensorflow as tf
```
Docstring style (lines 16-24) doesn't hanging indent.   Example stolen from off the style guide:

"""A one line summary of the module or program, terminated by a period.

Leave one blank line.  The rest of this docstring should contain an
overall description of the module or program.  Optionally, it may also
contain a brief description of exported classes and functions and/or usage
examples.

  Typical usage example:

  foo = ClassFoo()
  bar = foo.FunctionBar()
"""

alan-krumholz replied:
will fix

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -25,11 +25,11 @@
 import math
```
Don't forget to re-lint this file, too.

alan-krumholz replied:
When doing the previous solution (in which this one is based) we decided not to use gpylint (as it is not publicly available) so we used public pylint instead and the code was passing all those checks

alan-krumholz replied:
will do

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/task.py:
```diff
@@ -50,45 +50,40 @@ def initialise_params():
     args_parser.add_argument(
```
Also needs re-linting

alan-krumholz replied:
will do

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -44,15 +42,8 @@ def parse_csv(record):
-        distribution_defaults + weather_defaults
-    column_names = [TARGET_COLUMN, 'date', 'day',
-                    'hour'] + distribution_cols + weather_cols
+    header_def = [[''], [0], [0]]
```
nit: "def" can be interpreted both as "default" and "definition".  I also don't understand the meaning of "header" in this name.  How about just "record_defaults" or 'column_defaults"?

wes-turner replied:
nit: I found the name "parse_csv" a little misleading because I think of a CSV as an entire file instead of an encoding of a single record.  Is this function name standard?

alan-krumholz replied:
will change name to column_defaults

I named it parse_csv as I'm using same wording TF uses for their function to parse lists of comma separated strings : tf.decode_csv

wes-turner replied:
My reading of it is that tf.decode_csv, CSV is a set of records, whereas in parse_csv a CSV is a single record.  Please resolve this without making any edits if this is pedantic/unhelpful rather than pedantic/helpful.

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -44,15 +42,8 @@ def parse_csv(record):
-                    'hour'] + distribution_cols + weather_cols
+    header_def = [[''], [0], [0]]
+    column_names = ['id', 'depth', TARGET_COLUMN]
     columns = tf.decode_csv(record, record_defaults=header_def)
```
Why does it make sense to have default 'depth' and 'salt'?  If it does make sense, could you please explain in a comment?

alan-krumholz replied:
record_defaults are a requirement in tf.decode_csv to deal with missing values when loading data. These are not used if there aren't missing values but tf needs these values (it also use them to infer the type [e.g. int, float, str])

wes-turner replied:
I think I still don't understand.  I expect it to be better to either throw an exception or drop the record if data is missing.  Why have you chosen to not mark all columns as required?

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -69,12 +60,40 @@ def get_features_target_tuple(features):
     return features, target
 
+def load_image(image_path):
+    """Loads and process an image.
```
nit: Loads and processes an image.

wes-turner replied:
"process" tells me that something more than encoding is possibly going on.  Here, I think it just encodes it?

alan-krumholz replied:
will fix typo

It encodes it and also resizes it.

wes-turner replied:
nit: Encodes and resizes an image is a good docstring.  But it's fine as it is, too -- I recognize that this is a dumb utility function and doesn't warrant much attention.

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -69,12 +60,40 @@ def get_features_target_tuple(features):
+    """Loads and process an image.
+
+    Args:
+        image_path: Path to image.
```
From https://engdoc.corp.google.com/eng/doc/devguide/py/style/index.md#383-functions-and-methods:
"The description should include required type(s) if the code does not contain a corresponding type annotation."

alan-krumholz replied:
what would a "corresponding type annotation" look like? Type (string) can be inferred by how is used

wes-turner replied:
"String with a path to an image" is exactly what I'd meant.  The docs for tf.read_file, for example, describe its filename input as "A Tensor of type string." (an alternate type could be a PosixPath or something).

alan-krumholz replied:
will fix

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -69,12 +60,40 @@ def get_features_target_tuple(features):
+    """
+    image_string = tf.read_file(image_path)
+    image_decoded = tf.image.decode_png(image_string,channels=3)
+    image_resized = tf.image.resize_images(image_decoded, [224, 224])
```
Maybe make 224 a module-level constant?

alan-krumholz replied:
will do

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -69,12 +60,40 @@ def get_features_target_tuple(features):
+    return image_resized  
+
+def process_features(features, image_path):
+    """Returns processed features.
```
Could you please add a short description of what the processing is?

alan-krumholz replied:
will do

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -69,12 +60,40 @@ def get_features_target_tuple(features):
+
+    Args:
+        features: Dictionary with all columns.
+        image_path: Path to image folder.
```
Are there any assumptions to consider about the contents of this folder?

alan-krumholz replied:
not really as tf.read_file will work fine with either local system paths or gs:// paths

wes-turner replied:
I'd meant assumptions about the folder's directory structure.  I think maybe one of them, for example, is that there's a "<id>.png" for every <id> in the training CSV.  I'm typing this comment thinking about the user experience for someone trying to set up their own experiment (complete with their own inputs).

alan-krumholz replied:
will change docstring to explain better

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -69,12 +60,40 @@ def get_features_target_tuple(features):
+    """Returns processed features.
+
+    Args:
+        features: Dictionary with all columns.
```
What are "all" columns?  Is "image" a column?  (this is a dumb nit - feel free to ignore if it's pedantic for no purpose)

alan-krumholz replied:
will rephrase

wes-turner replied:
Setting aside our conversation about how to annotate types, the return value here could be described a little more thoroughly.  I'd like something like "a features dict augmented or transformed by this function's processing."  It's tougher to know exactly what to write when one of the goals is trying to keep the code general and easy to copy.

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -69,12 +60,40 @@ def get_features_target_tuple(features):
     target = features.pop(TARGET_COLUMN, None)
     return features, target
 
+def load_image(image_path):
```
What do you think about marking all these helper functions as "private"?  (load_image -> _load_image, etc.)

alan-krumholz replied:
I'm not sure as I need to pass this function to tf.map_fn (as well use it internally):
tf.map_fn(load_image ..)
load_image(image_path

wes-turner replied:
The "private" underscore for defining the public/private interface of a module:

PEP8: https://www.python.org/dev/peps/pep-0008
_single_leading_underscore: weak "internal use" indicator. 

Internal stye-guide: https://engdoc.corp.google.com/eng/doc/devguide/py/style/index.md#3164-guidelines-derived-from-guidos-recommendations
"_lower_with_under" for module-internal functions.

If this wasn't TF-internsive, I think the guidance is clear.  But I haven't written many TF utilities, and so don't know if standards and idioms are different there.

Either way, this isn't the biggest deal and whatever you want to do is fine.

wes-turner replied:
I've given this some more thought.  If you don't want to go through and denote each modules public/private interface with underscores, let's at least talk about it before just resolving this item.  Additionally, I'd talk about this item more if you'd like to do so just to make sure we're both well-informed and understand each other.

alan-krumholz replied:
will fix

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -69,12 +60,40 @@ def get_features_target_tuple(features):
+        image_path: Path to image folder.
         shuffle: Boolean flag specifying if data should be shuffled.
         batch_size: Number of records to be read at a time.
         num_epochs: Number of times to go through all of the records.
```
nit: useed -> used

alan-krumholz replied:
will fix

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -93,10 +112,15 @@ def _input_fn():
+        dataset = dataset.map(
+            get_features_target_tuple,
+            num_parallel_calls=num_threads)
+        dataset = dataset.map(lambda features, target: (
```
Does this read easier if features are processed before the target is popped off the dict?

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -107,33 +131,48 @@ def _input_fn():
 
     Returns:
-        A `ServingInputReceiver`.
+        Serving function.
```
Does this need description?

alan-krumholz replied:
will fix

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -107,33 +131,48 @@ def _input_fn():
+        features = parse_csv(csv_row)
+        features, _ = get_features_target_tuple(features)
+        features['image'] = tf.map_fn(
+            load_image, image_path + features['id'] + '.png',
```
I don't know if this is worth DRYing, but it's the sort of thing that sometimes makes me uncomfortable writing.  Just wanted to drop the note in case you want to do something with it.  I certainly don't think you need to.

alan-krumholz replied:
what do you mean by DRYing?

wes-turner replied:
DRY = don't repeat yourself, and DRYing is refactoring code to make multiple uses of a thing use just one single implementation.

In this case it's the image_directory_path + features['id'] + '.png' -> image_path.

alan-krumholz replied:
will fix

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -107,33 +131,48 @@ def _input_fn():
-    features, _ = get_features_target_tuple(features)
+        features = parse_csv(csv_row)
+        features, _ = get_features_target_tuple(features)
+        features['image'] = tf.map_fn(
```
I am highly inexpert when it comes to TensorFlow.  Why did you choose map_fn for what looks like simple function application?

alan-krumholz replied:
map_fn helps you apply the function in parallel (and get the output in the desired shape).
This is equivalent to the dataset.map I used for training. tf.map_fn helps me do the same at inference for several records at the same time.

wes-turner replied:
I think of a dataset as like a collection of records.  Dataset.map(f) maps f over each record.

tf.map_fn(f, c) similarly maps a function over a collection.  And as per your comment, I think I understand how it helps TF apply the function in parallel.

But here you're using it even when collection contains just a single item.  Is there parallelism that I'm not seeing?

alan-krumholz replied:
The size of the tensor is not known:
csv_row = tf.placeholder(
            shape=[None],
            dtype=tf.string
        )
Note shape [None]^

That means that when we publish this model people can call it with how many records as they want (or fit in memory) and we'll process all in parallel (CMLE batch prediction product always calls this function with 50 examples every time. I think they hardcoded that number on their side)

wes-turner replied:
What I'm hearing is that features is a Tensor-collection of dicts and that features['id'] is a Tensor-collection of strings?

Then image_path + features['id'] + IMAGE_TYPE is a Tensor-collection of strings, too?

alan-krumholz replied:
features['id']  is a tensor collection of strings at that level, but if you do tf.map_fn(features['id']) (or dataset.map()) it applies a specific function to each element of that collection (separately but in parallel) and returns a collection with all elements transformed by that function.

wes-turner replied:
Awesome, awesome.  Thank you for walking me through that.

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -73,51 +73,40 @@ def create_regressor(config, parameters):
     Returns:
```
Putting a TODO into this docstring sounds nice.  I'm also not sure, reading the docstring, what is/shouldn't be necessary and what the more-desirable code would look like.

wes-turner replied:
Also, since CL/187793590 has been committed, the text of the docstring could change a little.

wes-turner replied:
(this comment is meant to be attached to "forward_key_to_export")

alan-krumholz replied:
is CL/187793590 in production/public? if so for which version of TF?

Brian made me get rid of all TODOS last time :) (I had one there)

wes-turner replied:
Bleh -- sorry for the conflicting guidance.  Do you remember Brian's rationale for removing TODOs?

That CL was submitted back in March.  Seems awfully likely it's live, but I don't know of any way of tying a file's Piper version to the released TF version, but I could go through GitHub to find this change.

I think 2 things need to be done, here.  The first is to describe why this is a workaround and what should be done when that CL hits live (This comment should not have been submitted as written back when Lak wrote it).  The second is to do check if it's hit live and, if it has, either do that thing or note that we're no longer waiting on the CL but just the work to do what's described in thing 1.

alan-krumholz replied:
Brian said no complete solution should be published with a TODO as it means the solution wasn't ready... (or something along those lines)

I'm asking Lak, if it is live we can delete the whole function :)

I'll let you know

wes-turner replied:
Thanks! And thanks for the context.

TODOs are referenced a number of places internally as "The characters TODO in a comment are used to call out code that is temporary, a short-term solution, or good-enough but not perfect. When present, it is in the format"

More than you ever wanted to read about TODOs is at go/tott-486. It also says, "a TODO highlights an existing or potential issue, points toward future direction, or suggests opportunities for improvements. A TODO allows readers to understand the problems with the existing code and how it can be improved."

While I think I understand Brian's concern about publishing solutions that look unfinished, I think his conclusion is different from how the company approaches TODOs (and how I like to use them, too).

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -73,51 +73,40 @@ def create_regressor(config, parameters):
-        mean = pickle.load(f)
-    with file_io.FileIO(parameters.std_path, mode='r') as f:
-        std = pickle.load(f)
+    mean = np.float32(parameters.depth_mean)
```
That these values need to be hardcoded is weird and seems like an opportunity for error.  Is there another alternative?

alan-krumholz replied:
they are not hardcoded. You can pass them as parameters when running the script (only the defaults are hard coded if parameters are missing when running the script)

wes-turner replied:
"hardcoded" was a poor choice of words.  My reading of this is that mean and std_dev are information that's used at training time but need to be explicitly passed here at inference.  If the case, it means that the caller has to remember facts about training.  That seems like an easy place for future bugs to emerge from.

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -73,51 +73,40 @@ def create_regressor(config, parameters):
-        mean = pickle.load(f)
-    with file_io.FileIO(parameters.std_path, mode='r') as f:
-        std = pickle.load(f)
+    mean = np.float32(parameters.depth_mean)
```
nit: could you please use depth_mean and depth_std as variable names?

alan-krumholz replied:
will do

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -73,51 +73,40 @@ def create_regressor(config, parameters):
-    # Evaluation metric.
-    def mean_absolute_error(labels, predictions):
-        """Creates mean absolute error metric.
+    def estimator_metrics(labels, predictions):
```
Why is this function nested inside create_regressor?  I don't see any variables being closed.

alan-krumholz replied:
because this function is only used to be passed to an estimator when creating one it shouldn't be used for anything else. They do the same in the CMLE official solution templates.

wes-turner replied:
:-( They are not following the style guide.

https://engdoc.corp.google.com/eng/doc/devguide/py/style/index.md#264-decision

"Avoid nested functions or classes except when closing over a local value. Do not nest a function just to hide it from users of a module. Instead, prefix its name with an _ at the module level so that it can still be accessed by tests.

alan-krumholz replied:
will fix

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -73,51 +73,40 @@ def create_regressor(config, parameters):
 
         Returns:
-            A dictionary with the evaluation metric
+            A dictionary with the evaluation metrics
```
Would anyone ever look to this docstring to see which metrics are being evaluated, or are they only ever going to care about the return type?

alan-krumholz replied:
I believe the latter (ideally they should be able to reuse this code without having to modify all doc strings)

wes-turner replied:
(y)

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -73,51 +73,40 @@ def create_regressor(config, parameters):
+        pred_class = predictions['class_ids']
+        return {
+            'accuracy': tf.metrics.accuracy(labels,  pred_class),
+            'auc': tf.metrics.auc(labels, pred_logistic),
```
Maybe call this auc_roc and make the curve explicit so as to make clear the difference with the metric below?

wes-turner replied:
Still outstanding.  If you think the code is better without, this is ok to resolve without changing anything.

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -73,51 +73,40 @@ def create_regressor(config, parameters):
+        Metrics are used to evaluate the model.
 
         Args:
             labels: Evaluation true labels.
```
Is it worth documenting what labels and predictions are?  The style guide says that at least type should be mentioned.  I might go so far as to say that columns are a part of any type.  What do you think is useful?

alan-krumholz replied:
We tried to make the code/doc strings as generic as possible so it is super easy for someone to take it and use it as a starting point for their similar (but different) solution without having to change too much code (or too many doc strings)

wes-turner replied:
That's totally fair.  Thanks for the explanation!

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -73,51 +73,40 @@ def create_regressor(config, parameters):
+        """Creates metrics for Estimator.
 
-        Metric is used to evaluate the model.
+        Metrics are used to evaluate the model.
```
There's some opportunity for describing what the function does, and it's also ok if it's self-evident enough that no comment is needed.  But this comment doesn't communicate much.

alan-krumholz replied:
I expanded the explanation to give more insights.

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -73,51 +73,40 @@ def create_regressor(config, parameters):
-        pred_values = predictions['predictions']
-        return {'mae': tf.metrics.mean_absolute_error(
-            labels, pred_values)}
+        pred_logistic = predictions['logistic']
```
This might not be the right place to comment, but it's non-obvious to me what 'logistic' by its name is supposed to represent.

alan-krumholz replied:
'logistic' is what TF API expects as an input to give you back probabilities for the classification predictions. This is documented in the TF API.

wes-turner replied:
Bleh.  That's an unfortunate name for (I'm assuming) "Tensor of weights to which a logistic function has been applied."  But it sounds like from what you're saying, it's not even that but is "Tensor of weights to which a softmax has been applied."  Is that about right?

Also, and continuing my own education, where is predictions generated?  Is 'logistic' a magic string?

alan-krumholz replied:
See the second code block:
https://www.tensorflow.org/api_docs/python/tf/contrib/estimator/add_metrics

'logistic' is  a key in a predictions dictionary produced by the canned TF Estimator

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -127,14 +116,15 @@ def mean_absolute_error(labels, predictions):
         layer = h_units[-1]
 
-    estimator = tf.estimator.DNNRegressor(
+    estimator = tf.estimator.DNNClassifier(
```
There are still places in this file that refer to the estimator as a regressor.

alan-krumholz replied:
will fix

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -73,51 +73,40 @@ def create_regressor(config, parameters):
     Returns:
         A configured and ready to use `tf.estimator.DNNRegressor`
```
What makes an estimator "configured"?  The docstrings in this module haven't, in general, been very helpful to me.  They feel a little like they were written so that the file would have docstrings.

If it's because I don't have the TensorFlow experience that most people will have (and this file is assuming), that's ok, too.  In that case, I think I recommend just using the single-line short form.

alan-krumholz replied:
configured = "estimator instance with specified columns, architecture, metrics, optimizer, etc"

Initially my docstrings were much bigger/descriptive but Brian complained a lot about them as he said I was just describing the logic of the code and doc strings shouldn't be that.

I'm open to change this to anything you suggest.

alan-krumholz commented:
@wes-turner I've addressed most of your questions/suggestions. Please let me know if you have any followup comments.
Hope you are enjoying your time there! Thanks so much for your help!

alan-krumholz commented:
@wes-turner  I've addressed most of your questions/suggestions (for real this time :)) 
Please let me know if you have any followup comments.

Thanks so much for your help!

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -71,53 +70,44 @@ def create_regressor(config, parameters):
+        """Creates metrics for Estimator.
 
-        Metric is used to evaluate the model.
+        Metrics defined here can be used to evaluate the model (on evaluation
```
I like this docstring.

alan-krumholz replied:
:)

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -93,10 +119,17 @@ def _input_fn():
+            get_features_target_tuple,
+            num_parallel_calls=num_threads)
+        dataset = dataset.map(
+            lambda features, target: (
```
https://engdoc.corp.google.com/eng/doc/devguide/py/style/index.md#210-lambda-functions says lambdas are ok for "oneliners".  It's super-obnoxious to refactor this, and I understand why you implemented it this way.

I'm not sure what I'd do, myself.  Probably, I'd do all I can to get it to fit on one line.  I think in this case that it's also fine to do nothing.

I thought I'd made this comment last review, but I don't see it now.  If you swap the order of this map and the map before, does the whole thing read clearer?  Then, maybe, you don't need to do any sort of acrobatics.

if you choose to do something such that gpylint still complains, you'll either need to disable that warning now, or if/when this repo is ported to //third_party, you'll need to go back and disable the warning then.

alan-krumholz replied:
This comes from the template:
https://github.com/GoogleCloudPlatform/cloudml-samples/blob/master/cloudml-template/template/trainer/input.py#L227

Not sure what you mean by swapping the order of the maps?  this one needs to happen after  and should not operate in labels (in case labels are not available during production  inference)

wes-turner replied:
get_features_target_tuple: features -> (features-target, target)
process_features: features x image_path -> features+image

I claim that these functions are commutable.  If so, then we can perform the following rewrite:

dataset = dataset.map(get_features_target_tuple, num_parallel_calls=num_threads)
dataset = dataset.map(lambda features, target: (process_features(features, image_path), target), num_parallel_calls=num_threads)

->

dataset = dataset.map(lambda features: process_features(features, image_path), num_parallel_calls=num_threads)
dataset = dataset.map(get_features_target_tuple, num_parallel_calls=num_threads)

wes-turner replied:
I didn't understand your comment about labels.  But I also don't want to ignore it.  If it's still relevant, could you please explain what you meant?

alan-krumholz replied:
I understand now. I'll change it. I think the template has it that way in case you need some logic to produce the label that'll be invalid if process_features is ran before.  but in this case this won't happen so I'll change as suggested.

wes-turner replied:
(y)

It's tough to reason about what the code should look like when one's scope is larger than just the application/library it's being written for.

alan-krumholz commented:
@wes-turner, I think I addressed all but the one of the CL. Please let me know if you still have nay open concerns (apart form that one) and I will try to wrap them all soon (including the CL one)

Thanks so much for your help!

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -44,16 +44,9 @@ def parse_csv(record):
     Returns:
```
@wes-turner
wes-turner 6 days ago  Collaborator
nit: "def" can be interpreted both as "default" and "definition". I also don't understand the meaning of "header" in this name. How about just "record_defaults" or 'column_defaults"?

  @wes-turner
wes-turner 6 days ago  Collaborator
nit: I found the name "parse_csv" a little misleading because I think of a CSV as an entire file instead of an encoding of a single record. Is this function name standard?

  @alan-krumholz
alan-krumholz 5 days ago  Owner
will change name to column_defaults

I named it parse_csv as I'm using same wording TF uses for their function to parse lists of comma separated strings : tf.decode_csv

  @wes-turner
wes-turner 21 hours ago  Collaborator
My reading of it is that tf.decode_csv, CSV is a set of records, whereas in parse_csv a CSV is a single record. Please resolve this without making any edits if this is pedantic/unhelpful rather than pedantic/helpful.

alan-krumholz replied:
actually parse_csv will work with either a set of records or a set of a single record. 
in _input_fn() is used on sets of 1  records but in csv_serving_input_fn() is used in a set of undefined  size.

wes-turner replied:
Ooh -- very cool.

This should be noted in the Args part of the docstring.

alan-krumholz replied:
will fix

wes-turner replied:
This is dumb, but for the docs is there a technical description for the type that 'record' can be?  The TF docs for decode_csv give, for record

> A Tensor of type string. Each string is a record/row in the csv and all records should have the same format.

I think it'd be ok to adapt that (or copy/paste it, since it's Google copyright)

alan-krumholz replied:
will fix

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -44,16 +44,9 @@ def parse_csv(record):
-    column_names = [TARGET_COLUMN, 'date', 'day',
-                    'hour'] + distribution_cols + weather_cols
-    columns = tf.decode_csv(record, record_defaults=header_def)
+    column_defaults = [[''], [0], [0]]
```
@wes-turner
wes-turner 6 days ago  Collaborator
Why does it make sense to have default 'depth' and 'salt'? If it does make sense, could you please explain in a comment?

  @alan-krumholz
alan-krumholz 5 days ago  Owner
record_defaults are a requirement in tf.decode_csv to deal with missing values when loading data. These are not used if there aren't missing values but tf needs these values (it also use them to infer the type [e.g. int, float, str])

  @wes-turner
wes-turner 21 hours ago  Collaborator
I think I still don't understand. I expect it to be better to either throw an exception or drop the record if data is missing. Why have you chosen to not mark all columns as required?

alan-krumholz replied:
I see what you mean now. It is supposed to take empty Tensor objects to denote as required. Let me play with this to make sure I make it work as needed. Will update here once I have it working

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -70,17 +63,51 @@ def get_features_target_tuple(features):
 
 
-def generate_input_fn(file_path, shuffle, batch_size, num_epochs):
+def _load_image(image_path):
```
@wes-turner
wes-turner 6 days ago  Collaborator
nit: Loads and processes an image.

  @wes-turner
wes-turner 6 days ago  Collaborator
"process" tells me that something more than encoding is possibly going on. Here, I think it just encodes it?

  @alan-krumholz
alan-krumholz 5 days ago  Owner
will fix typo

It encodes it and also resizes it.

  @wes-turner
wes-turner 21 hours ago  Collaborator
nit: Encodes and resizes an image is a good docstring. But it's fine as it is, too -- I recognize that this is a dumb utility function and doesn't warrant much attention.

alan-krumholz replied:
will fix

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -70,17 +63,51 @@ def get_features_target_tuple(features):
+    """Loads and processes an image.
+
+    Args:
+        image_path: String with a path to an image.
```
wes-turner 6 days ago  Collaborator
From https://engdoc.corp.google.com/eng/doc/devguide/py/style/index.md#383-functions-and-methods:
"The description should include required type(s) if the code does not contain a corresponding type annotation."

  @alan-krumholz
alan-krumholz 5 days ago  Owner
what would a "corresponding type annotation" look like? Type (string) can be inferred by how is used

  @wes-turner
wes-turner 21 hours ago  Collaborator
"String with a path to an image" is exactly what I'd meant. The docs for tf.read_file, for example, describe its filename input as "A Tensor of type string." (an alternate type could be a PosixPath or something).

  @alan-krumholz
alan-krumholz 11 hours ago  Owner
will fix

wes-turner replied:
You've added a lot of type information in your docstrings.  Thanks!  How would you like to treat the places that you haven't changed from the file you copied this from, but that still don't have them?

I, myself, would like them to be added -- both for consistency's sake in how it reads, and because it's easier to review to a standard in the future when that standard is consistently applied.  But that's easy for me to ask for because I'm not the person who'd actually have to do it.

What are your thoughts?

alan-krumholz replied:
right, didn't notice these path arguments were all over the place. I think I've fixed them all now

wes-turner replied:
Thanks for doing that.  This sort of work helps the work be a reference solution to internal engineers, too.

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -70,17 +63,51 @@ def get_features_target_tuple(features):
+
+    Args:
+        features: Dictionary with data features.
+        image_path: String with path to the folder containing training images.
```
wes-turner 6 days ago  Collaborator
Are there any assumptions to consider about the contents of this folder?

  @alan-krumholz
alan-krumholz 5 days ago  Owner
not really as tf.read_file will work fine with either local system paths or gs:// paths

  @wes-turner
wes-turner 21 hours ago  • 
 Collaborator
I'd meant assumptions about the folder's directory structure. I think maybe one of them, for example, is that there's a ".png" for every in the training CSV. I'm typing this comment thinking about the user experience for someone trying to set up their own experiment (complete with their own inputs).

  @alan-krumholz
alan-krumholz 11 hours ago  Owner
will change docstring to explain better

wes-turner replied:
Is what the contents of this directory have to look like documented somewhere else?

alan-krumholz replied:
added some description to the function to explain this

wes-turner replied:
I like it.

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -70,17 +63,51 @@ def get_features_target_tuple(features):
+        image_path: String with path to the folder containing training images.
+
+    Returns:
+        processed features.
```
@wes-turner
wes-turner 6 days ago  Collaborator
What are "all" columns? Is "image" a column? (this is a dumb nit - feel free to ignore if it's pedantic for no purpose)

  @alan-krumholz
alan-krumholz 5 days ago  Owner
will rephrase

  @wes-turner
wes-turner 20 hours ago  Collaborator
Setting aside our conversation about how to annotate types, the return value here could be described a little more thoroughly. I'd like something like "a features dict augmented or transformed by this function's processing." It's tougher to know exactly what to write when one of the goals is trying to keep the code general and easy to copy.

alan-krumholz replied:
will add

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -70,17 +63,51 @@ def get_features_target_tuple(features):
 
 
-def generate_input_fn(file_path, shuffle, batch_size, num_epochs):
+def _load_image(image_path):
```
wes-turner 6 days ago  Collaborator
What do you think about marking all these helper functions as "private"? (load_image -> _load_image, etc.)

  @alan-krumholz
alan-krumholz 5 days ago  Owner
I'm not sure as I need to pass this function to tf.map_fn (as well use it internally):
tf.map_fn(load_image ..)
load_image(image_path

  @wes-turner
wes-turner 21 hours ago  • 
 Collaborator
The "private" underscore for defining the public/private interface of a module:

PEP8: https://www.python.org/dev/peps/pep-0008
_single_leading_underscore: weak "internal use" indicator.

Internal stye-guide: https://engdoc.corp.google.com/eng/doc/devguide/py/style/index.md#3164-guidelines-derived-from-guidos-recommendations
"_lower_with_under" for module-internal functions.

If this wasn't TF-internsive, I think the guidance is clear. But I haven't written many TF utilities, and so don't know if standards and idioms are different there.

Either way, this isn't the biggest deal and whatever you want to do is fine.

  @wes-turner
wes-turner 20 hours ago  Collaborator
I've given this some more thought. If you don't want to go through and denote each modules public/private interface with underscores, let's at least talk about it before just resolving this item. Additionally, I'd talk about this item more if you'd like to do so just to make sure we're both well-informed and understand each other.

  @alan-krumholz
alan-krumholz 11 hours ago  Owner
will fix

wes-turner replied:
I also left a similar comment about type annotation.  What are your thoughts about going through the whole CL and adjusting function names to fit the PEP8/Google-style recommendation?

alan-krumholz replied:
Changed all function names across the solution

wes-turner replied:
Thanks!  I imagine that being a small headache.

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -108,32 +142,48 @@ def _input_fn():
+        features = parse_csv(csv_row)
+        features, _ = get_features_target_tuple(features)
+        features['image'] = tf.map_fn(
+            _load_image, image_path + features['id'] + IMAGE_TYPE,
```
wes-turner 6 days ago  Collaborator
I don't know if this is worth DRYing, but it's the sort of thing that sometimes makes me uncomfortable writing. Just wanted to drop the note in case you want to do something with it. I certainly don't think you need to.

  @alan-krumholz
alan-krumholz 5 days ago  Owner
what do you mean by DRYing?

  @wes-turner
wes-turner 21 hours ago  Collaborator
DRY = don't repeat yourself, and DRYing is refactoring code to make multiple uses of a thing use just one single implementation.

In this case it's the image_directory_path + features['id'] + '.png' -> image_path.

  @alan-krumholz
alan-krumholz 11 hours ago  Owner
will fix

wes-turner replied:
Though this is DRYer, the sort of DRYing I'd had in mind was factoring the path creation into a function
f: directory_path x id -> image_path

It's totally ok to say that this is overkill.  I think I'd do it that way, myself, but recognize it's a matter of taste.  I'd originally left this comment not to mandate one solution or another, but call your attention to the repetition in case it was something that had gotten lost for you and that you'd want to rewrite, yourself.

alan-krumholz replied:
got it. will add

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -50,74 +48,67 @@ def forward_key_to_export(estimator):
+        A configured and ready to use `tf.estimator.DNNClassifier`
+    """
+    # Mean and Standard Deviation Constants for normalization.
+    depth_mean = np.float32(parameters.depth_mean)
```
@wes-turner
wes-turner 6 days ago  Collaborator
That these values need to be hardcoded is weird and seems like an opportunity for error. Is there another alternative?

  @alan-krumholz
alan-krumholz 5 days ago  Owner
they are not hardcoded. You can pass them as parameters when running the script (only the defaults are hard coded if parameters are missing when running the script)

  @wes-turner
wes-turner 20 hours ago  Collaborator
"hardcoded" was a poor choice of words. My reading of this is that mean and std_dev are information that's used at training time but need to be explicitly passed here at inference. If the case, it means that the caller has to remember facts about training. That seems like an easy place for future bugs to emerge from.

alan-krumholz replied:
what do you suggest to improve this?

wes-turner replied:
I really don't know.  Is this a standard, "Whelp, our tooling ain't great but there's nothing we can do" type of problem?  Not writing much TF, I don't what problems everybody agrees are problems, but are hard and nobody's come up with any solutions to.

alan-krumholz replied:
In my previous solution I had a data preparation script that would do all the complicated  data munging including generating means and stds for my 180 features.

https://github.com/GoogleCloudPlatform/professional-services/blob/master/examples/cloudml-energy-price-forecasting/data_preparation/data_prep.py

In this solution we don't need any complicated data munging that's why I didn't create such script (I think it'll be an overkill to have one just to generate these two numbers don't you think? that's why I decided just to let the user input them with the job )

wes-turner replied:
I agree with you that a script like that is overkill.  There should be a way to embed per-record preprocessing directives into a served model as a part of packaging it up after training.  ... I don't suppose anything like that exists?

alan-krumholz replied:
There is a TF extension for DataFlow but running a DataFlow job for this will be even a bigger overkill than a script.

I can create a script like in the last solution for these if you think that is the right way forward.

wes-turner replied:
I like your solution as written.  I'll follow up with you about it next time we talk to continue my own education and understand the toolset better.  Thanks for the chat, here.

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -50,74 +48,67 @@ def forward_key_to_export(estimator):
+    pred_class = predictions['class_ids']
+    return {
+        'accuracy': tf.metrics.accuracy(labels, pred_class),
+        'auc': tf.metrics.auc(labels, pred_logistic),
```
@wes-turner
wes-turner 6 days ago  Collaborator
Maybe call this auc_roc and make the curve explicit so as to make clear the difference with the metric below?

  @wes-turner
wes-turner 21 hours ago  Collaborator
Still outstanding. If you think the code is better without, this is ok to resolve without changing anything.

alan-krumholz replied:
will do

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -50,74 +48,67 @@ def forward_key_to_export(estimator):
-        A configured and ready to use `tf.estimator.DNNRegressor`
+        A dictionary with the evaluation metrics
     """
+    pred_logistic = predictions['logistic']
```
@wes-turner
wes-turner 6 days ago  Collaborator
This might not be the right place to comment, but it's non-obvious to me what 'logistic' by its name is supposed to represent.

  @alan-krumholz
alan-krumholz 5 days ago  Owner
'logistic' is what TF API expects as an input to give you back probabilities for the classification predictions. This is documented in the TF API.

  @wes-turner
wes-turner 20 hours ago  Collaborator
Bleh. That's an unfortunate name for (I'm assuming) "Tensor of weights to which a logistic function has been applied." But it sounds like from what you're saying, it's not even that but is "Tensor of weights to which a softmax has been applied." Is that about right?

Also, and continuing my own education, where is predictions generated? Is 'logistic' a magic string?

  @alan-krumholz
alan-krumholz 11 hours ago  Owner
See the second code block:
https://www.tensorflow.org/api_docs/python/tf/contrib/estimator/add_metrics

'logistic' is a key in a predictions dictionary produced by the canned TF Estimator

wes-turner replied:
Ah!  Makes sense about where the key, value pair comes from.  Thanks!

wes-turner replied:
I'm still grumpy about its name.

wes-turner replied:
(not grumpy at you)

wes-turner replied:
Also, also, tf.estimator.DNNClassifier.predict does not have a very helpful "Yields" section in its docstring.  It seems like that's the place where the contents of the predictions dict should be documented.

alan-krumholz replied:
Agree, I think TF docs haven't't kept up with the rapid product development

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -50,74 +48,67 @@ def forward_key_to_export(estimator):
 
-    feature_cols = [hour, day] + weather + distribution
+    Returns:
+        A configured and ready to use `tf.estimator.DNNClassifier`
```
@wes-turner
wes-turner 6 days ago  Collaborator
What makes an estimator "configured"? The docstrings in this module haven't, in general, been very helpful to me. They feel a little like they were written so that the file would have docstrings.

If it's because I don't have the TensorFlow experience that most people will have (and this file is assuming), that's ok, too. In that case, I think I recommend just using the single-line short form.

  @alan-krumholz
alan-krumholz 5 days ago  Owner
configured = "estimator instance with specified columns, architecture, metrics, optimizer, etc"

Initially my docstrings were much bigger/descriptive but Brian complained a lot about them as he said I was just describing the logic of the code and doc strings shouldn't be that.

I'm open to change this to anything you suggest.

wes-turner replied:
> describing the logic of the code and doc strings shouldn't be that.

Agreed about that.  The style guide says:

> A docstring should describe the function's calling syntax and its semantics, not its implementation.

The line between the two is pretty fine when the logic is simple.  What I'd wanted here was something about semantics -- what it meant to be configured.  I want to know what to expect from/about the returned object itself.  What's been done to it and what do I still need to do?  For example, does depth need to be normalized?

alan-krumholz replied:
changed the description hope you like it better.

alan-krumholz commented:
@wes-turner 
Thanks for copying all again. Not sure why they disappeared :(
I think I've addressed all your comments except two:

1) Getting rid of Lak's hack (he confirmed this shouldn't be needed as of TF 1.8)
2) See if I can configure the tf.dataset so all columns are required.

I'll be working on these two later today and will update you once they are addressed.

Thank you!

alan-krumholz commented:
@wes-turner 
The following two have been addressed now:
1) Getting rid of Lak's hack (he confirmed this shouldn't be needed as of TF 1.8)
2) See if I can configure the tf.dataset so all columns are required.

Please let me know if there is anything else missing.
Thank you!

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -1,39 +1,39 @@
 
-  inputs.get_train_spec(parameters)
-  inputs.get_eval_spec(parameters)
+inputs.get_train_spec(parameters)
```
I don't think this usage is quite "typical" -- the return values aren't even saved.  Could you please either:

* add a couple more lines of code for what one is supposed to do with the artifacts generated by these functions
or 
* describe in English what the important functions are and how to use them?

alan-krumholz replied:
will fix

wes-turner replied:
That looks great!

alan-krumholz replied:
good!

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -44,20 +44,16 @@ def parse_csv(record):
-    column_names = [TARGET_COLUMN, 'date', 'day',
-                    'hour'] + distribution_cols + weather_cols
-    columns = tf.decode_csv(record, record_defaults=header_def)
+    column_defaults = [
```
Thanks!  The docstring is slightly out of date, now.  I think it's ok to leave it as-is, and this is for sure just a nit, but I think two possibilities you could consider are either:

* removing the part about defaults
* mentioning that in this particular solution, there are no defaults and all the values are required

alan-krumholz replied:
will fix

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -70,17 +66,68 @@ def get_features_target_tuple(features):
+    return image_resized
+
+
+def _create_image_path(image_path, image_id):
```
Purely FYI, and what you choose to do is just your own personal style, but it's ok to omit the "returns" annotation in the docstrings for these tiny utility functions when the docstring starts with the word "returns":

https://engdoc.corp.google.com/eng/doc/devguide/py/style/index.md#383-functions-and-methods
>  It may also be omitted if the docstring starts with Returns (or Yields) (e.g. """Returns row from Bigtable as a tuple of strings.""") and the opening sentence is sufficient to describe return value.

wes-turner replied:
Not an endorsement, just informational.

alan-krumholz replied:
will fix

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/inputs.py:
```diff
@@ -1,39 +1,39 @@
 #!/usr/bin/env python
-#Copyright 2018 Google LLC
+# Copyright 2018 Google LLC
```
Please re-lint this file before submitting.  It breaks a couple dumb syntax rules.

alan-krumholz replied:
will fix

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/model.py:
```diff
@@ -71,53 +63,21 @@ def create_regressor(config, parameters):
 
     Returns:
-        A configured and ready to use `tf.estimator.DNNRegressor`
+        `tf.estimator.DNNClassifier` with specified features and architecture.
```
I for sure do like this new return section of the docstring.  Thanks!

alan-krumholz replied:
Good! :)

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/task.py:
```diff
@@ -127,10 +122,16 @@ def initialise_params():
+    args_parser.add_argument(
+        '--tf_hub_module',
+        help='TF Hub module to use for images.',
+        default='https://tfhub.dev/google/imagenet/mobilenet_v2_140_224/feature_vector/2',
```
This line is too long and the linter complains.  My favorite fixes are either:
* break up the string into two (maybe after the ".dev") and put the second string on a new line
* disable the line-too-long check for this line

I think you can maybe also stick the whole string onto its own line, too, and make it fit that way?  But that solution would be a pretty ugly one.

In the end, I'm happy with anything you choose to do here so the linter doesn't complain.

alan-krumholz replied:
will fix

wes-turner replied:
:-) I like your solution.

alan-krumholz commented:
@wes-turner 
I think I've now addressed them all :)
Please let me know your thought.

Thank you!

wes-turner commented on machine-learning/solutions/seismic_salt_detection/trainer/task.py:
```diff
@@ -50,45 +50,40 @@ def initialise_params():
-        default=0.2,
+        '--depth_mean',
+        help='Mean depth on training data.',
+        default=507.904,
```
It seems dangerous to me to have a default mean and standard deviation.  I want this library's user to know that there's normalization that he or she has to configure.

What are your thoughts?

alan-krumholz replied:
I'm adding these to the README scripts so people know about them without going through the code

alan-krumholz commented:
@wes-turner 
I think I fixed all your last comments.
Please give it a last review and let me know if you are OK with it

Thank you!
